feat(dpg): support `logicalPath` of lro

archerzz opened the pull request:
# Description

- pass in `logicalPath` info from emitter
- in generator, generate convert lambda to fetch the return value from `logicalPath`
- add test case

resolve #3837


# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

archerzz commented on test/TestProjects/Authoring-TypeSpec/Authoring-TypeSpec.tsp:
```diff
@@ -225,7 +225,7 @@ model SwapDeploymentsParams {
 
 interface Projects {
   @doc("Creates a new project or updates an existing one.")
-  createOrUpdate is LongRunningResourceCreateOrUpdate<Project>;
```
The reason to replace `LongRunningResourceCreateOrUpdate` with `LongRunningResourceCreateOrReplace` is:
- there is an issue in `LongRunningResourceCreateOrUpdate`: https://github.com/Azure/typespec-azure/issues/3726 we could handle patch case but that will make our logic more complex
- LRO patch seems an invalid case, we should remove or deprecate it. See https://github.com/Azure/typespec-azure/issues/3257

archerzz commented:
Regen preview: https://github.com/Azure/azure-sdk-for-net/pull/39323

It also rolls back the workaround for `OpenAI`: [Azure/azure-sdk-for-net@`046ae62` (#39323)](https://github.com/Azure/azure-sdk-for-net/pull/39323/commits/046ae62b09df34719fa71aa6ea694a17fd055ebd)

AlexanderSher commented on src/TypeSpec.Extension/Emitter.Csharp/src/lib/operation.ts:
```diff
@@ -400,6 +404,13 @@ export function loadOperation(
+                    // https://github.com/Azure/typespec-azure/issues/3725
+                    (metadata.envelopeResult != metadata.logicalResult &&
+                    op.verb === "post"
+                        ? "result" // actually `result` is the only allowed path for now
```
Can this be more than just one property?

archerzz replied:
@AlexanderSher Currently it should be only one property. I used to have a snippet of codes which will parse the `logicalPath` and convert into a property expression like `A.B.C`. After discussing with people more familiar with the current implementation of LRO in typespec, I abandoned the codes to keep it simple. We can add the logic back if needed, it's not complex.

AlexanderSher commented on src/AutoRest.CSharp/Common/Input/InputTypes/OperationResponse.cs:
```diff
@@ -3,10 +3,30 @@
+internal record OperationResponse(IReadOnlyList<int> StatusCodes, InputType? BodyType, BodyMediaType BodyMediaType, IReadOnlyList<OperationResponseHeader> Headers, bool IsErrorResponse, string? ResultPath)
 {
-    public OperationResponse() : this(StatusCodes: Array.Empty<int>(), BodyType: null, BodyMediaType: BodyMediaType.None, Headers: Array.Empty<OperationResponseHeader>(), IsErrorResponse: false) { }
+    public OperationResponse() : this(StatusCodes: Array.Empty<int>(), BodyType: null, BodyMediaType: BodyMediaType.None, Headers: Array.Empty<OperationResponseHeader>(), IsErrorResponse: false, ResultPath: null) { }
```
`ResultPath` is meaningful only in LRO. Can we move this property to the `OperationLongRunning`?

archerzz replied:
done

archerzz commented on src/AutoRest.CSharp/LowLevel/Generation/DpgClientWriter.cs:
```diff
@@ -360,12 +360,26 @@ private void WriteConvenienceLroMethod(LowLevelClientMethod clientMethod, Conven
+
+            var bodyType = longRunning.FinalResponse.BodyType!;
+            return @$"r => {{
+                {bodyType.Name} rawResponse = {bodyType.Name}.FromResponse(r);
```
Using `System.TextJson` can also eliminate the generation of intermediate models.

archerzz commented:
List of changes since last review:
- Move `ReturnType` and `ResultPath` from `OperationResponse` to `OperationLongRunning`.
- Use `System.Text.Json` to fetch the result JSON data, instead of using raw response model to deserialize.
- adopt `feature/v3` changes

AlexanderSher commented on test/CadlRanchProjects/azure/core/lro/standard/src/Generated/StandardClient.cs:
```diff
@@ -284,7 +288,11 @@ public virtual Operation<ExportedUser> Export(WaitUntil waitUntil, string name,
+            return ProtocolOperationHelpers.Convert(response, r =>
+            {
+                var resultJsonElement = System.Text.Json.JsonDocument.Parse(r.Content).RootElement.GetProperty("result");
+                return ExportedUser.DeserializeExportedUser(resultJsonElement);
```
This two-line delegate looks really ugly. Maybe generate it as a local function?

AlexanderSher replied:
Somethind like:
```cs
ExportedUser DeserializeResult(Response response)
{
    using var document = JsonDocument.Parse(r.Content);
    return ExportedUser.DeserializeExportedUser(document.RootElement.GetProperty("result"));
}
return ProtocolOperationHelpers.Convert(response, DeserializeResult, ClientDiagnostics, "StandardClient.Export");
```

archerzz commented on test/CadlRanchProjects/azure/core/lro/standard/src/Generated/StandardClient.cs:
```diff
@@ -437,5 +438,11 @@ internal static RequestContext FromCancellationToken(CancellationToken cancellat
         private static ResponseClassifier _responseClassifier202;
         private static ResponseClassifier ResponseClassifier202 => _responseClassifier202 ??= new StatusCodeClassifier(stackalloc ushort[] { 202 });
+
+        private ExportedUser FetchExportedUserFromResourceOperationStatusUserExportedUserError(Response response)
```
Conversion method is named after pattern `Fetch{ReturnType}From{RawResponseType`. For this test case, the name is pretty long, because `RawResponseType` is a typespec core template.

chunyu3 commented on src/TypeSpec.Extension/Emitter.Csharp/src/lib/operation.ts:
```diff
@@ -401,7 +405,14 @@ export function loadOperation(
+            } as OperationResponse,
+            ResultPath:
+                metadata.logicalPath ??
+                // https://github.com/Azure/typespec-azure/issues/3725
```
Add a TODO here to inform we will remove this workaround after issue 3725 is resoved.

chunyu3 commented on src/TypeSpec.Extension/Emitter.Csharp/test/TestProjects/lro/lro.tsp:
```diff
@@ -47,5 +47,5 @@ model Pet {
 interface Projects {
     @doc("Creates a new pet or updates an existing one.")
-    createOrUpdate is LongRunningResourceCreateOrUpdate<Pet>;
+    createOrUpdate is LongRunningResourceCreateOrReplace<Pet>;
```
Why do we update the test here? Have we decided not to support `LongRunningResourceCreateOrUpdate` ?

archerzz replied:
See my comments at top in `Authoring-Typespec`. After discussing with Weidong, I decided to not support `LongRunningResourceCreateOrUpdate` until it's clarified. No real service is using it. So we're fine.

chunyu3 commented on test/CadlRanchProjects/azure/core/lro/standard/src/Generated/StandardClient.cs:
```diff
@@ -265,7 +266,7 @@ public virtual async Task<Operation<ExportedUser>> ExportAsync(WaitUntil waitUnt
             RequestContext context = FromCancellationToken(cancellationToken);
             Operation<BinaryData> response = await ExportAsync(waitUntil, name, format, context).ConfigureAwait(false);
-            return ProtocolOperationHelpers.Convert(response, ExportedUser.FromResponse, ClientDiagnostics, "StandardClient.Export");
+            return ProtocolOperationHelpers.Convert(response, FetchExportedUserFromResourceOperationStatusUserExportedUserError, ClientDiagnostics, "StandardClient.Export");
```
Can we just name `FetchExportedUserFromResponse`?

archerzz replied:
There is possibility that `ExportedUsers` could be results from different LRO with different raw responses. It would be better to name after `Fetch{ReturnType}From{RawResponseType}` to avoid naming conflicts.

chunyu3 replied:
So the main difference is the `logicPath` among different raw responses. In that way, can we just write an common ResultFetch function `FetchExportedUserFromRequpons(Response response, string logicPath=null)`, in that way, we may not generate many Fetch methods.
But in order to support this, we need to update a lot in `ProtocolOperationHelpers`. It can be an refine in future.
Current it is ok to write multiple Fetch method, one for one Lro.

chunyu3 commented on src/AutoRest.CSharp/LowLevel/Output/OperationMethodChainBuilder.cs:
```diff
@@ -162,6 +162,14 @@ private void BuildSamples(LowLevelClientMethod method, List<DpgOperationSample>
 
+        private LongRunningResultRetrievalMethod? GetLongRunningResultRetrievalMethod(OperationLongRunning? longRunning)
+        {
+            if (longRunning is null || longRunning.ResultPath is null)
```
We may simplify this as `if (longRunning?.ResultPath is null)`

chunyu3 commented on src/TypeSpec.Extension/Emitter.Csharp/src/lib/operation.ts:
```diff
@@ -388,7 +388,11 @@ export function loadOperation(
-                metadata.logicalResult
+                // TODO: we should check `logicalPath` or other ways to determine body type,
+                // after https://github.com/Azure/typespec-azure/issues/3725 is fixed
+                op.verb === "post"
```
How about `op.verb !== "post"` operation? Can a `put` Lro operation has `envelopeResult`?

archerzz replied:
> How about `op.verb !== "post"` operation? Can a `put` Lro operation has `envelopeResult`?

From the spec, `logicalPath` is for `post` (e.g. RPC), at least for now. That's why I'm using `=== "post"`.

archerzz replied:
The better way is to check `logicalPath` regardless of HTTP verb. But there is an issue here.

lirenhe commented on src/AutoRest.CSharp/Common/Input/InputTypes/OperationLongRunning.cs:
```diff
@@ -2,10 +2,29 @@
+    public OperationLongRunning() : this(FinalStateVia: OperationFinalStateVia.Location, FinalResponse: new OperationResponse(), null) { }
+
+    /// <summary>
+    /// Meaninngful return type of the long running operation.
```
typo?

archerzz commented:
Changes since last review:
- fix typo and add todo
- reverse null check for better readability
- merge `feature/v3`

Preview PR: https://github.com/Azure/azure-sdk-for-net/pull/39439
